### ISSUE TITLE
XAML Islands: error C3779 - 'winrt::impl::consume_Windows_UI_Xaml_Controls_Primitives_ISelector<D>::SelectedIndex': a function that returns 'auto' cannot be used before it is defined

### ISSUE THREAD

**Issue by pbatard** [microsoft/WindowsAppSDK#2031]
### Describe the bug

Following the [_"Host a standard WinRT XAML control in a C++ desktop (Win32) app"_ tutorial](https://docs.microsoft.com/en-us/windows/apps/desktop/modernize/host-standard-control-with-xaml-islands-cpp) and trying to work with a ComboBox control, one gets the following compilation error in VS2022 after adding some entries and then trying to use the `SelectedIndex()` method set the default item:

```
error C3779: 'winrt::impl::consume_Windows_UI_Xaml_Controls_Primitives_ISelector<D>::SelectedIndex': a function that returns 'auto' cannot be used before it is defined
```

### Steps to reproduce the bug

1. In VS2022 create a project and code according to the guide at https://docs.microsoft.com/en-us/windows/apps/desktop/modernize/host-standard-control-with-xaml-islands-cpp
2. Replace the  `Windows::UI::Xaml::Controls::TextBlock tb` related code with the following:  
```C++
    // Create a ComboBox with 2 items and select the first one as default
    Windows::UI::Xaml::Controls::ComboBox cb;
    cb.FontSize(16);
    cb.Items().Append(winrt::box_value(L"Item 1"));
    cb.Items().Append(winrt::box_value(L"Item 2"));
    cb.SelectedIndex(0);
    xamlContainer.Children().Append(cb);
```
3. Try to compile. You will get the error:  
```
error C3779: 'winrt::impl::consume_Windows_UI_Xaml_Controls_Primitives_ISelector<D>::SelectedIndex': a function that returns 'auto' cannot be used before it is defined
1>        with
1>        [
1>            D=winrt::Windows::UI::Xaml::Controls::ComboBox
1>        ]
```

### Expected behavior

I'm hoping that I'm missing something about the subtleties of C++ or, more likely, how combobox items might required to be bound to some data set, but I really don't get how the ComboBox class can be defined so that `cb.FontSize(16)` invocation will work, but not `cb.SelectedIndex(0)`, regardless of whether items have been added or not (since its whole point is to act on the selection at runtime). 

The fact that there should exist an index `0` and an index `1` after `Items().Append()` has been invoked, and that these entries are most definitely available when the control is displayed, makes it very puzzling for a newcomer to understand why the implementation of the XAML Islands ComboBox wants to prevent the use of `cb.SelectedIndex()` before we get a chance to invoke it at runtime...

### Screenshots

Of course the issue is that, unless you can call on `cb.SelectedIndex()` then the default control will display like this:
![Image1](https://user-images.githubusercontent.com/1206968/151296513-bff39abf-1cb8-4518-a2cc-b945788c502f.png)

instead of the desired:
![Image2](https://user-images.githubusercontent.com/1206968/151296550-71310260-c42a-4044-9056-c52e15596d58.png)

### NuGet package version

1.0.0

### Packaging type

Unpackaged

### Windows version

Windows 11 version 21H2 (22000)

### IDE

Visual Studio 2022

**Comment by BenJKuhn** [comment]
This is a common usage error that people encounter with C++ /WinRT. It's a tradeoff that optimizes build times by requiring the developer to include headers for each namespace used. Outside of the namespaces you explicitly include, C++ /WinRT only includes enough of the definition to enable the types you requested to compile, but not enough to be usable. WinRT types are highly interconnected in some places, so the transitive closure of includes would lead to very undesirable build performance. 

Raymond Chen blogged about this in more detail here:
https://devblogs.microsoft.com/oldnewthing/20190530-00/?p=102529

**Comment by riverar** [comment]
Add `#include <winrt/Windows.UI.Xaml.Controls.Primitives.h>`

**Comment by riverar** [comment]
@BenJKuhn @oldnewthing In hopes of encouraging another look at this, I'd like to share the data point that this was also my first encounter with the reported error and neither the blog post or this error were particularly helpful.

**Comment by pbatard** [comment]
If I may chime in about my troubleshooting process, so that maybe it'll help you understand where a user, who isn't that familiar with either C++ or WinRT (I've mostly been programming desktop win32 apps in pure C, using good old `.rc` files as well as straight HWND manipulation APIs) is coming from, I did find some information on the internet that mentioned that the aforementioned error message was usually due to a missing header.

So I first tried to see if there were a `<winrt/Windows.ui.xaml.controls.combobox.h>` or `<winrt/Windows.ui.xaml.controls.iselector.h>` (since the error pointed to the `struct consume_Windows_UI_Xaml_Controls_Primitives_ISelector` implementation) I could include, but obviously these don't exist.

The closest I got to solving my issue was adding a `using namespace Windows::UI::Xaml::Controls::Primitives;` statement, but that's where I failed to push it one step further and see if there also was a `<winrt/Windows.ui.xaml.controls.primitives.h>` to include.

A this stage, after trying to find examples of XAML Island code calling `SelectedIndex()` on GitHub, my thought process was _"Well, XAML Islands are quite new, and it's possible that there are some virtual functions defined, that either I'm supposed to add some plumbing for in the code, or that have yet to be added by the WinRT/C++ team"_, and I therefore logged this issue to get some advice.

Still, I sure I appreciate the amount of thought that is being given on trying to clarify the cause of this error for developers who, like myself, are probably running into these kind of hurdles due to an obvious lack of familiarity with some elements and language concepts that other developers might be able to navigate more easily...
